tests/csapi/thread_notifications_test.go

```diff
@@ -0,0 +1,196 @@
+//
+// 1. Send a series of messages, some of which are in threads.
+// 2. Send combinations of threaded and unthreaded receipts.
+// 3. Ensure the notification counts are updated appropriately.
+//
+// This sends four messages as alice creating a DAG somewhat like:
+//
+// A<--B<--C
+// ^
+// +---D
+//
+// Where C and D generate highlight notifications.
+//
+// Notification counts and receipts are handled by bob.
+func TestThreadedReceipts(t *testing.T) {
+	runtime.SkipIf(t, runtime.Dendrite) // not supported
+	deployment := Deploy(t, b.BlueprintOneToOneRoom)
+	defer deployment.Destroy(t)
+
+	// Create a room with alice and bob.
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+	bob := deployment.Client(t, "hs1", "@bob:hs1")
+
+	roomID := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat"})
+	bob.JoinRoom(t, roomID, nil)
+
+	token := bob.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(bob.UserID, roomID))
+
+	// Send an initial message as alice.
+	firstEventID := alice.SendEventSynced(t, roomID, b.Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "Hello world!",
+		},
+	})
+
+	// Create a thread from the above message and send both two messages in it,
+	// the second of which is a mention (causing a highlight).
+	firstThreadEventID := alice.SendEventSynced(t, roomID, b.Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "Start thread!",
+			"m.relates_to": map[string]interface{}{
+				"event_id": firstEventID,
+				"rel_type": "m.thread",
+			},
+		},
+	})
+	alice.SendEventSynced(t, roomID, b.Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    fmt.Sprintf("Thread response %s!", bob.UserID),
+			"m.relates_to": map[string]interface{}{
+				"event_id": firstEventID,
+				"rel_type": "m.thread",
+			},
+		},
+	})
+
+	// Send an additional unthreaded message, which is a mention (causing a highlight).
+	secondEventID := alice.SendEventSynced(t, roomID, b.Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    fmt.Sprintf("Hello %s!", bob.UserID),
+		},
+	})
+
+	// A filter to get thread notifications.
+	threadFilter := `{"room":{"timeline":{"unread_thread_notifications":true}}}`
+
+	// Check the unthreaded and threaded counts, which should include all previously
+	// sent messages.
+	bob.MustSyncUntil(
+		t, client.SyncReq{Since: token},
+		client.SyncTimelineHas(roomID, func(r gjson.Result) bool {
+			return r.Get("event_id").Str == secondEventID
+		}),
+		syncHasUnreadNotifs(roomID, func(r gjson.Result, t gjson.Result) bool {
+			return r.Get("highlight_count").Num == 2 && r.Get("notification_count").Num == 4 && !t.Exists()
+		}),
+	)
+	bob.MustSyncUntil(
+		t,
+		client.SyncReq{Since: token, Filter: threadFilter}, client.SyncTimelineHas(roomID, func(r gjson.Result) bool {
```

Comment:
Suggest starting `SyncTimelineHas` on a new line. (Give me `black` for golang. Surprised gofmt doesn't just enforce this for you---I thought go was all about opinionated tooling?)

Reply:
I thought `gofmt` was more opinionated than it was...